[Windows][UWP] VSync Detection Fixes and Improvements

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

When using the "Sync to playback" mode, the "missed vsync" counter of the player debug info (ctrl-shift-O) constantly increases for both win64 and uwp on win64 (most likely on xbox as well). The higher the display refresh rate, the faster the counter increases.

After investigation with VS "Concurrency Visualizer", the largest contributor to the missed vsyncs is dxgi's multithreading synchronization. The  vsync detection loop runs on a different thread than rendering and a few functions get blocked by the Present() calls on the main thread, which block from the completion of rendering until vsync. As a result, every second vsync is missed.

It took a while and a lot of testing to figure things out and create the PR as there were many unexpected side effects...

One commit per change, descriptions in the commit messages.

The change with the most impact is the removal of the sleep of half a vsync period, which was necessary with Windows 7 but not with later versions of Windows. It caused the output release before WaitForVBlank to wait for Present of main thread to complete, and missed the vblank. With UWP, in addition, the lack of API to increase timer resolution did not allow sleeping accurately enough.

The last commit uses the dxgi output maintained by windowing code rather than retrieving it from dxgi for every frame. It's not 100% necessary to do after the half-vsync sleep removal, but still helps reduce resource usage and provides more safety for nVidia + combined HDR/refresh rate switch (see testing).

Testing for refresh rate change was kept for safety when HDR status is switched, because there is an intermediate state right after HDR switch when the refresh rate reverts to desktop rate before it's changed again to the wanted refresh rate. There is nothing preventing the video clock code from sampling the refresh rate exactly at this transition time, and to retrieve an incorrect value. It happened from time to time during testing.
Could be removed after improving events around the HDR switch. Throwing the events OnLostDisplay/OnResetDisplay is enough to cause the video clock to sample the refresh rate again.
Until then, optimized the code a bit to be less costly per frame.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Missed vsync counter increases quickly in the player debug info (ctrl-shift-O) in Windows 10 x64 with native and UWP builds (most likely on xbox as well).

That doesn't seem right, although it's not clear that there is much impact because the video clock seems to interpolate from the last good vsync.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 10 64 bits, native and UWP version. Windows 8 64 bits.
Windows 7: not available but should still be OK since the half-vsync sleep was not changed.
Xbox: testing required, I don't have the HW.

Tried AMD, Intel (multiple generations) and nVidia, with/without refresh rate and HDR switching > missed counter usually doesn't increase after initial rendering initializations. Stressing the system can still cause some misses but that seems difficult to avoid completely.

One weird problem noticed (but pre-existing so no action needed in PR): with nVidia and HDR + refresh rate switch, the vsync detection works fine, but the main thread / rendering still presents at the original desktop refresh rate speed instead of the new refresh rate > not normal. May explain some jerkyness/stutter/...
This is clearly visible with the Concurrency Visualizer tool of VS. No problem with AMD in same scenario.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Theoretically slightly better synchronization, but not clear as it seems that the video reference clock interpolates the good vsyncs to hide the missed vsyncs.

## Screenshots (if appropriate):
With the PR, low and steady missed count
![image](https://github.com/xbmc/xbmc/assets/489377/2f1a518e-056b-4621-9635-61b666bd04e5)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[Windows] videosync: retrieve refresh rate only on display config change.

The previous condition was true for every frame and retrieved the current refresh rate too often.

IsCurrent() also reacts to display changes other than refresh rate but display changes are rare.
IsCurrent() takes about 1/10th of the time of refresh rate retrieval and mostly saves time

CrystalP committed:
[Windows] Handle short WaitForVBlank

This happens when closing the lid of a laptop, during refresh rate/HDR status change and for errors.
Before attempting to detect another vblank, sleep to avoid a busy loop.
Since the user can't see the output, no need to attempt to maintain perfect vsync with an expensive
accurate alternative source such as spinning on performance counters.
It's better to not send updates to the video reference clock and let it interpolate from the last
known good timestamp)

CrystalP committed:
[Windows] Retrieve the screen refresh rate without using the swap chain

The video clock runs on its own thread.
Previous implementation with DeviceResources::GetDisplayMode() used a swap chain method that sometimes
gets stalled by an internal dxgi lock held by the main thread calling Present(). The additional
information is not even useful here => skip

CrystalP committed:
[Windows] const-ify system clock frequency

CrystalP committed:
[Windows] Add GetOutput() alternative, using cached information first.

Windowing events already maintain an up to date dxgi output member. Make it track the output
description as well, and use those members from the vsync tracking rather than spending resources
retrieving the current dxgi output every frame.

Other benefits:
- avoids dxgi synchronization with main thread Present() for GetContainingOutput() and output->GetDesc().
- avoids GetContainingOutput() errors when the swapchain doesn't own the output, for example with
WARP device and sometimes with multi-screen UWP desktop that can't catch some events due to UWP limitations.

thexai commented:
For now tested on Intel NUC8 (Windows 11) and everything seems fine, tomorrow I will test on Xbox Series S

thexai commented on xbmc/windowing/windows/VideoSyncD3D.cpp:
```diff
@@ -87,20 +122,29 @@ void CVideoSyncD3D::Run(CEvent& stopEvent)
+#if defined(TARGET_WINDOWS_DESKTOP)
+    // Half-vsync sleep is not needed for UWP or Win > 7 (no freeze with back to back WaitForVBlank())
+    // It's actually a problem for UWP because of the lack of high resolution timers so exclude.
+    // Keep for desktop Windows for now, there is no need to break compatibility with Windows 7 on purpose.
```
Windows 7 is not longer supported on Omega then if "is not needed for UWP or Win > 7" this code block can be removed?

CrystalP replied:
Yes it can but will definitely break Windows 7 according to previous comments and other sources on the internet.

There is a difference between not supporting Win 7 and knowingly breaking it. I'm not sure if we're doing the former or the latter for v21.
It's not really in the way here (except for UWP with insufficient timer accuracy), but removing would probably allow a smaller fix. I'd have to run some tests.

thexai replied:
From https://github.com/xbmc/xbmc/pull/17255 minimal Windows version required is 8.1. 

Is not  even possible install v21 on Windows 7 (installer blocks it).

thexai commented on xbmc/windowing/windows/VideoSyncD3D.cpp:
```diff
@@ -120,14 +164,33 @@ void CVideoSyncD3D::Cleanup()
+  if (hdmiInfo) // Xbox only
+  {
+    auto currentMode = hdmiInfo.GetCurrentDisplayMode();
+    m_fps = currentMode.RefreshRate();
```
`m_fps` is float and `currentMode.RefreshRate();` double --> needs static_cast

T:\KODI\kodi\xbmc\windowing\windows\VideoSyncD3D.cpp(188,36): warning C4244: '=': conversion from 'double' to 'float', possible loss of data

thexai commented:
Runtime tested on Xbox: first confirmed issue is present before and resolved after. All is right :) 

Note that "missed" increases to 2 - 3 at beginning but it does not increase further during playback.

CrystalP commented:
Thanks for the testing, I'll review after removing the sleep, other simplifications may be possible.

CrystalP committed:
[Windows] Remove unnecessary sleep for half vsync

Windows 7 is not supported anymore in v21 and the sleep was necessary only for Windows 7 as back to
back WaitForVBlank calls would otherwise freeze rendering.

Sleep is not needed anymore with Windows 8 and later, and detrimental with UWP, which doesn't have an
API to increase timer precision and frequently overslept.

CrystalP commented:
Added the cast from your comment and retested exhaustively. A few changes are not 100% necessary but kept in the end for safety so I ended up only removing the Windows 7 sleep and reducing a bit the sleep when WaitForVBlank returns quickly, because it also happens sometimes when switching HDR.

The initial "Missed" seem to come from Windows blocking some functions when switching HDR / refresh rate but I didn't dig too much. It would make sense for us to suspend that counter while switching but it's not really worth the effort...
